client/allocrunner/alloc_runner.go

```diff
@@ -873,17 +855,15 @@ func (ar *allocRunner) Shutdown() {
 		ar.logger.Trace("shutting down")
 
 		// Shutdown tasks gracefully if they were run
-		if ar.runnersLaunched {
-			wg := sync.WaitGroup{}
-			for _, tr := range ar.tasks {
-				wg.Add(1)
-				go func(tr *taskrunner.TaskRunner) {
-					tr.Shutdown()
-					wg.Done()
-				}(tr)
-			}
-			wg.Wait()
+		wg := sync.WaitGroup{}
+		for _, tr := range ar.tasks {
+			wg.Add(1)
+			go func(tr *taskrunner.TaskRunner) {
+				tr.Shutdown()
```

Comment:
how would this respond to a partial shutdown - not sure if we'd be able to recover then? (e.g if a single task runner Shutdown were to panic, and nomad got restarted?)

Reply:
A panic from tr.Shutdown would crash Nomad. Since Shutdown's don't affect task or alloc state everything should start back up and reattach properly.

The only change here was always calling tr.Shutdown since tr.Run is always called on TaskRunners. AR.Shutdown is called on every AR in `Client.allocs` and there's no opportunity for Nomad to exit or error between adding ARs to that  map and calling AR.Run on them. So it should not be possible to get into a situation where Shutdown will block forever waiting on a TR.Run that never starts.

Reply:
👍